aws/lambda/oss_ci_aws_ce_tracking/lambda_function.py

```diff
@@ -0,0 +1,334 @@
+                    "Values": ["Amazon Elastic Compute Cloud - Compute"],
+                }
+            },
+        )
+
+    def _process_raw_ce_data(self, resp_reults: List[Dict[str, Any]]):
+        filtered = copy.deepcopy(resp_reults)
+        # filter NoInstanceType since it's normally related to other usages
+        for res in filtered:
+            # filted Groups with keys `NoInstanceType``
+            filtered_groups = [
+                group
+                for group in res.get("Groups", [])
+                if "NoInstanceType" not in group.get("Keys", [])
+            ]
+            res["Groups"] = filtered_groups
+
+        # flatten the aws results in to a record that is ready for clickhouse schema conversion
+        res = []
+        for ts_record in filtered:
+            falttened_list = self.flatten_ts_record(ts_record)
+            res.extend(falttened_list)
+        return res
+
+    def flatten_ts_record(self, record: Dict[str, Any]) -> List[Dict[str, Any]]:
+        results = []
+        start = record.get("TimePeriod", {}).get("Start")
+        end = record.get("TimePeriod", {}).get("End")
+        for group in record["Groups"]:
+            new_group = {}
+            new_group["Start"] = start
+            new_group["End"] = end
+            new_group["Keys"] = group.get("Keys", [])
+            usage_quantity = group.get("Metrics", {}).get("UsageQuantity", {})
+            new_group["Amount"] = usage_quantity.get("Amount", 0)
+            new_group["Unit"] = usage_quantity.get("Unit", 0)
+            results.append(new_group)
+        return results
+
+    def to_db_schema(
+        self, record: Dict[str, Any], record_type: str, granularity: str
+    ) -> Optional[Dict[str, Any]]:
+        keys = record.get("Keys", [])
+        startTime = record.get("Start", "").replace("Z", "+00:00")
+        now = datetime.now(timezone.utc).isoformat()
+        if len(keys) < 2:
+            logger.warning(
+                f"Expected two keys from Record, but got {len(record)} keys:{keys}, skipping the record"
+            )
+            return None
+        return {
+            "created": now,
+            "type": record_type,
+            "granularity": granularity,
+            "time": startTime,
+            "instance_type": keys[0],
+            "usage_type": keys[1],
+            "unit": record.get("Unit", ""),
+            "value": record.get("Amount", 0),
+            "extra_info": {},
+            "tags": [],
+        }
+
+    def start(self, args: Optional[argparse.Namespace] = None):
+        """
+        Starts the processor
+        Parameters:
+        - args (Optional[argparse.Namespace]): Command-line arguments for local execution.
+        """
+
+        # set up time range for fetching data from AWS Cost Explorer
+        time_now = datetime.now(timezone.utc)
+        logger.info(f"Starting job at UTC time {time_now}")
+
+        end_time = datetime.now(timezone.utc).date() - timedelta(days=1)
```

Comment:
We can also use end_time=datetime.now(), and switch the table to ReplacingMergeTree, so that we have more recent data for the current day (not always complete, but will be replaced for the same primary keys)